servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/ReplayStrategies.java

```diff
@@ -0,0 +1,287 @@
+            if (ttl.isNegative()) {
+                throw new IllegalArgumentException("ttl: " + ttl + "(expected non-negative)");
+            }
+            if (maxItems <= 0) {
+                throw new IllegalArgumentException("maxItems: " + maxItems + "(expected >0)");
+            }
+            this.executor = requireNonNull(executor);
+            this.ttlNanos = ttl.toNanos();
+            this.maxItems = maxItems;
+            // SpSc, but needs iterator.
+            // accumulate is called on one thread (no concurrent access on this method).
+            // timerFire maybe called on another thread
+            items = new ConcurrentLinkedQueue<>();
+        }
+
+        @Override
+        public void accumulate(@Nullable final T t) {
+            long scheduleTimerNanos = -1;
+            final TimeStampSignal<T> signal = new TimeStampSignal<>(executor.currentTime(NANOSECONDS), t);
+            if (tryAcquireLock(queueLockUpdater, this)) {
+                for (;;) {
+                    final int qSize = queueSize;
+                    if (qSize < maxItems) {
+                        if (queueSizeUpdater.compareAndSet(this, qSize, qSize + 1)) {
+                            items.add(signal);
+                            if (qSize == 0) {
+                                scheduleTimerNanos = ttlNanos;
+                            }
+                            break;
+                        }
+                    } else if (queueSizeUpdater.compareAndSet(this, qSize, qSize)) {
+                        // Queue removal is only done while queueLock is acquired, so we don't need to worry about
+                        // the timer thread removing items concurrently.
+                        items.poll();
+                        items.add(signal);
+                        break;
+                    }
+                }
+                if (!releaseLock(queueLockUpdater, this)) {
+                    scheduleTimerNanos = tryDrainQueue();
+                }
+            } else {
+                queueSizeUpdater.incrementAndGet(this);
+                items.add(signal);
+                scheduleTimerNanos = tryDrainQueue();
+            }
+
+            if (scheduleTimerNanos >= 0) {
+                schedulerTimer(scheduleTimerNanos);
+            }
+        }
+
+        @Override
+        public void deliverAccumulation(final Consumer<T> consumer) {
+            int i = 0;
+            for (TimeStampSignal<T> timeStampSignal : items) {
+                consumer.accept(timeStampSignal.signal);
+                // The queue size maybe larger than maxItems if we weren't able to acquire the queueLock while adding.
+                // This is only a temporary condition while there is concurrent access between timer and accumulator.
+                // Guard against it here to preserve the invariant that we shouldn't deliver more than maxItems.
+                if (++i >= maxItems) {
+                    break;
+                }
```

Comment:
This feels like it can turn to an item leak.
```
maxItems = 1, items=[T1]
Thread1: timer fires and we take the queueLock, and suspend.
items=[T1]
Thread2: accumulate(T2), fail to take queueLock, adds T2 to items, and exits.
items=[T1, T2]
Thread2: new subscription, enters deliverAccumulation(..), emit T1, ++i == maxItems -> break; <-!! sees T1 but missed item T2.
```
I think this is intrinsically racy behavior: a program shouldn't be able to tell whether it received [T1, T2] from accumulation, a post-subscribe onNext calls, or a mix, so we should just deliver everything in the queue.

Reply:
great point. this is tricky condition to test test too.